src/metabase/driver/sql_jdbc/sync/describe_table.clj

```diff
@@ -0,0 +1,164 @@
+      (str "Invalid type: " special-type))
+    special-type))
+
+(defn- fallback-fields-metadata-from-select-query
+  "In some rare cases `:column_name` is blank (eg. SQLite's views with group by) fallback to sniffing the type from a
+  SELECT * query."
+  [driver ^Connection conn table-schema table-name]
+  (let [[sql & params] (common/simple-select-probe-query driver table-schema table-name {:select [:*]})]
+    (reify clojure.lang.IReduceInit
+      (reduce [_ rf init]
+        (with-open [stmt (common/prepare-statement driver conn sql params)
+                    rset (.executeQuery stmt)]
+          (reduce
+           rf
+           init
+           (let [metadata (.getMetaData rset)]
+             (eduction (map (fn [^Integer i]
+                              {:type_name   (.getColumnTypeName metadata i)
+                               :column_name (.getColumnName metadata i)}))
+                       (range 1 (inc (.getColumnCount metadata)))))))))))
+
+(defn- fields-metadata
+  "Returns reducible metadata for the Fields in a `table`."
+  [driver conn {schema :schema, table-name :name} & [^String db-name-or-nil]]
+  {:pre [(instance? Connection conn) (string? table-name)]}
+  (reify clojure.lang.IReduceInit
+    (reduce [_ rf init]
+      (with-open [rs (.getColumns (.getMetaData conn)
+                                  db-name-or-nil
+                                  (driver/escape-entity-name-for-metadata driver schema)
+                                  (driver/escape-entity-name-for-metadata driver table-name)
+                                  nil)]
+        ;; 1. Return all the Fields that come back from DatabaseMetaData that include type info.
+        ;;
+        ;; 2. Iff there are some Fields that don't have type info, concatenate
+        ;;    `fallback-fields-metadata-from-select-query`, which fetches the same Fields using a different method.
+        ;;
+        ;; 3. Filter out any duplicates between the two methods using `m/distinct-by`.
+        (let [missing-type-info? (volatile! false)
+              normal-fields      (eduction
+                                  (filter (fn [{type-name :type_name}]
+                                            (or (not (str/blank? type-name))
+                                                (do (vreset! missing-type-info? true)
+                                                    false))))
+                                  (jdbc/reducible-result-set rs {}))
+              fallback-fields    (reify clojure.lang.IReduceInit
+                                   (reduce [_ rf init]
+                                     (reduce
+                                      rf
+                                      init
+                                      (when @missing-type-info?
+                                        (fallback-fields-metadata-from-select-query driver conn schema table-name)))))]
+          (reduce
+           rf
+           init
+           (eduction
```

Comment:
A comment here warning to not rewrite any part of this as lazy would be good -- if I'm not mistaken, this only works reliably because transducers don't chunk, but doing something simple such as using concat on the input, rather than cat transducer, would break it. On a related note, wouldn't it be perhaps clearer/more robust to be somewhat less elegant, and just something like group-by normal fields based on whether it has `:type_name` and if `false` branch is not empty, mix in `fallback-fields`?

Reply:
I wasn't sure how to do that with transducers